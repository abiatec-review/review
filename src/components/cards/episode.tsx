import CharactersModal from '@components/modals/characters';
import TextRow from '@components/textRow';
import Episode from '@models/episode';
import {useSelector} from '@store';
import React, {useState} from 'react';
import {Pressable, StyleSheet} from 'react-native';

interface Props {
  episode: Episode;
}

function EpisodeCard({episode}: Props) {
  const [isModalShown, setIsModalShown] = useState(false);

  const toggleModal = () => setIsModalShown(!isModalShown);

  const characters = useSelector(({characterReducer}) =>
    characterReducer.characterList.filter(({url}) =>
      episode.characters.includes(url),
    ),
  );

  return (
    <>
      <Pressable style={styles.container} onPress={toggleModal}>
        <TextRow field="Name" data={episode.name} />
        <TextRow field="Air Date" data={episode.air_date} />
        <TextRow field="Code" data={episode.episode} />
        <TextRow field="Created" data={episode.created} />
      </Pressable>
      {characters.length > 0 && (
        <CharactersModal
          title="Characters"
          toggle={toggleModal}
          isShown={isModalShown}
          characters={characters}
        />
      )}
    </>
  );
}

export default React.memo(EpisodeCard);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    margin: 10,
    padding: 15,
    height: 140,
    borderRadius: 20,
    alignItems: 'center',
    backgroundColor: 'gray',
    justifyContent: 'space-between',
  },
});
