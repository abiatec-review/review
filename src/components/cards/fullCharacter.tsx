import TextRow from '@components/textRow';
import useOrientation from '@hooks/useOrientation';
import Character from '@models/character';
import React from 'react';
import {Image, StyleSheet, View} from 'react-native';

interface Props {
  character: Character;
}

function FullCharacterCard({character}: Props) {
  const {isPortrait} = useOrientation();
  const styles = isPortrait ? portraitStyles : landscapeStyles;

  return (
    <View style={styles.container}>
      <Image source={{uri: character.image}} style={styles.image} />
      <View style={styles.textBlock}>
        <TextRow field="Name" data={character.name} />
        <TextRow field="Status" data={character.status} />
        <TextRow field="Gender" data={character.gender} />
        <TextRow field="Origin" data={character.origin?.name} />
        <TextRow field="Created" data={character.created} />
      </View>
    </View>
  );
}

export default FullCharacterCard;

const baseStyles = StyleSheet.create({
  container: {
    padding: 15,
    borderRadius: 20,
    backgroundColor: 'gray',
  },
  image: {
    aspectRatio: 1,
    borderRadius: 20,
  },
  textBlock: {},
});

const portraitStyles = StyleSheet.create({
  ...baseStyles,
  image: {...baseStyles.image, marginBottom: 10},
});

const landscapeStyles = StyleSheet.create({
  ...baseStyles,
  container: {
    ...baseStyles.container,
    flexDirection: 'row',
    alignItems: 'center',
  },
  image: {
    ...baseStyles.image,
    flex: 0.4,
    marginRight: 10,
  },
  textBlock: {...baseStyles.textBlock, flex: 1},
});
