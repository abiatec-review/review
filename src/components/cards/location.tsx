import CharactersModal from '@components/modals/characters';
import TextRow from '@components/textRow';
import Location from '@models/location';
import {useSelector} from '@store';
import React, {useState} from 'react';
import {Pressable, StyleSheet} from 'react-native';

interface Props {
  location: Location;
}

function LocationCard({location}: Props) {
  const [isModalShown, setIsModalShown] = useState(false);

  const toggleModal = () => setIsModalShown(!isModalShown);

  const residents = useSelector(({characterReducer}) =>
    characterReducer.characterList.filter(({url}) =>
      location.residents.includes(url),
    ),
  );

  return (
    <>
      <Pressable style={styles.container} onPress={toggleModal}>
        <TextRow field="Name" data={location.name} />
        <TextRow field="Type" data={location.type} />
        <TextRow field="Dimension" data={location.dimension} />
        <TextRow field="Created" data={location.created} />
        <TextRow field="Residents" data={residents.length.toString()} />
      </Pressable>
      {residents.length > 0 && (
        <CharactersModal
          title="Residents"
          toggle={toggleModal}
          characters={residents}
          isShown={isModalShown}
        />
      )}
    </>
  );
}

export default (LocationCard);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    margin: 10,
    padding: 15,
    height: 170,
    borderRadius: 20,
    alignItems: 'center',
    backgroundColor: 'gray',
    justifyContent: "space-between",
  },
});
