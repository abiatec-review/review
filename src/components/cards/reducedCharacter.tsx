import CharacterDetailsModal from "@components/modals/characterDetails";
import Character from "@models/character";
import React, { useState } from "react";
import * as RN from "react-native";

interface Props {
  character: Character;
}

function ReducedCharacterCard({character}: Props) {
  const [isModalShown, setIsModalShown] = useState(false);

  const toggleModal = () => setIsModalShown(!isModalShown);

  return (
    <>
      <RN.Pressable style={styles.container} onPress={toggleModal}>
        <RN.Image source={{uri: character.image}} style={styles.image} />
        <RN.Text style={styles.text}>{character.name}</RN.Text>
      </RN.Pressable>
      <CharacterDetailsModal
        character={character}
        isShown={isModalShown}
        toggle={toggleModal}
      />
    </>
  );
}

export default ReducedCharacterCard;

const styles = RN.StyleSheet.create({
  container: {
    flex: 1,
    margin: 8,
    padding: 12,
    borderRadius: 20,
    flexDirection: 'column',
    backgroundColor: 'gray',
  },
  image: {
    width: '100%',
    aspectRatio: 1,
    borderRadius: 20,
    marginBottom: 10,
  },
  text: {
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
  },
});
