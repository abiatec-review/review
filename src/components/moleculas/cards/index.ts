export { CharacterCard } from "./character";
export { LocationCard } from "./location";
export { EpisodeCard } from "./episode";
