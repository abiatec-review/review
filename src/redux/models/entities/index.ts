export { Character } from "./character";
export { Location } from "./location";
export { Episode } from "./episode";
export { Pagination } from "./pagination";
export { ResultList } from "./pagination";
