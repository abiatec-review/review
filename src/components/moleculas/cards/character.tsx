import React, { useState } from "react";

import { Image, Pressable, StyleSheet, Text } from "react-native";

import { CharacterDetailsModal } from "@components/moleculas/modals";
import { Character } from "@redux/models/entities";
import { Color, FontSize, Indent, Radius } from "@utils";

interface Props {
  character: Character;
}

export function CharacterCard(props: Props) {
  const { character } = props;

  const [isModalShown, setIsModalShown] = useState(false);

  const toggleModal = () => setIsModalShown(!isModalShown);

  return (
    <>
      <Pressable style={styles.container} onPress={toggleModal}>
        <Image source={{ uri: character.image }} style={styles.image} />
        <Text style={styles.text}>{character.name}</Text>
      </Pressable>
      <CharacterDetailsModal character={character} isShown={isModalShown} toggle={toggleModal} />
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    margin: Indent.DEFAULT,
    padding: Indent.MEDIUM,
    flexDirection: "column",
    borderRadius: Radius.DEFAULT,
    backgroundColor: Color.TRANSPARENT_CYAN
  },
  image: {
    width: "100%",
    aspectRatio: 1,
    marginBottom: Indent.DEFAULT,
    borderRadius: Radius.DEFAULT
  },
  text: {
    fontWeight: "bold",
    color: Color.BLACK,
    textAlign: "center",
    fontSize: FontSize.DEFAULT
  }
});
