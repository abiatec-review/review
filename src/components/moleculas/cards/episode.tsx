import React, { useState } from "react";

import { Pressable, StyleSheet } from "react-native";

import { TextRow } from "@components/atoms";
import { CharactersModal } from "@components/moleculas/modals";
import { Episode } from "@redux/models/entities";
import { useSelector } from "@redux/store";
import { Colors, Indent, Radius } from "@utils";

interface Props {
  episode: Episode;
}

export function EpisodeCard(props: Props) {
  const { episode } = props;

  const [isModalShown, setIsModalShown] = useState(false);

  const toggleModal = () => setIsModalShown(!isModalShown);

  const characters = useSelector(({ character }) =>
    character.characterList.filter(({ url }) => episode.characters.includes(url))
  );

  return (
    <>
      <Pressable style={styles.container} onPress={toggleModal}>
        <TextRow field="Name" data={episode.name} />
        <TextRow field="Air Date" data={episode.air_date} />
        <TextRow field="Code" data={episode.episode} />
        <TextRow field="Created" data={episode.created} />
      </Pressable>
      {characters?.length > 0 && (
        <CharactersModal
          title="Characters"
          toggle={toggleModal}
          isShown={isModalShown}
          characters={characters}
        />
      )}
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    height: 140,
    alignItems: "center",
    padding: Indent.HUGE,
    margin: Indent.DEFAULT,
    borderRadius: Radius.DEFAULT,
    backgroundColor: Colors.GRAY,
    justifyContent: "space-between"
  }
});
