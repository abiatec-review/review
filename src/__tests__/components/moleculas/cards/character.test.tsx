import React from "react";

import { create } from "react-test-renderer";

import { CharacterCard } from "@components/moleculas/cards";
import { Character } from "@redux/models/entities";

jest.mock("@react-native-firebase/auth", () => {});
jest.mock("@react-native-firebase/storage", () => {});
jest.mock("react-native-image-picker", () => {});
jest.mock("react-native-vector-icons/MaterialIcons", () => {});
jest.mock("@hooks", () => ({ useOrientation: () => ({ isPortrait: true }) }));

const character: Character = {
  id: 1,
  name: "test",
  type: "test",
  gender: "test",
  status: "test",
  species: "test",
  origin: {
    name: "test",
    url: "test"
  },
  location: {
    name: "test",
    url: "test"
  },
  image: "test",
  episode: ["test"],
  url: "test",
  created: "test"
};

test("CharacterCard render", () => {
  const tree = create(<CharacterCard character={character} />).toJSON();
  expect(tree).toMatchSnapshot();
});
