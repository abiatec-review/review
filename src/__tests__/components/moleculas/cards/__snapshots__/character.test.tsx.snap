// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard render 1`] = `
Array [
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(175, 224, 209, 0.8)",
        "borderRadius": 20,
        "flex": 1,
        "flexDirection": "column",
        "margin": 10,
        "padding": 12,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "test",
        }
      }
      style={
        Object {
          "aspectRatio": 1,
          "borderRadius": 20,
          "marginBottom": 10,
          "width": "100%",
        }
      }
    />
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 18,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      test
    </Text>
  </View>,
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    transparent={true}
    visible={false}
  />,
]
`;
